### PR TITLE
recreate goals for 1.1rc1

### LINKED ISSUE #1098: recreate command should allow to keep a backup
im about to start to experiment with it, and i'd really like it better if by default it would backup and the same name case would only override the archive (the reason simply enabling to verify the result *before* loosing the source)

***

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/34667358-recreate-command-should-allow-to-keep-a-backup)

### LINKED ISSUE #1100: recreate should detect+skip already compressed chunks
> Detect and skip (unless --always-recompress) already recompressed chunks -- delayed until current PRs on borg.key APIs are decided #810 #789

(from #812)

***

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/34680260-recreate-should-detect-skip-already-compressed-chunks)

### PR DESCRIPTION
**Pull request by enkore**
Fixes #1098 #1100 #1146

### CODE REVIEW

File: src/borg/archive.py [hunk 1]
```diff
@@ -1387,21 +1389,29 @@ def process_item(self, archive, target, item):
...
         """Return new chunk ID list for 'item'."""
-        # TODO: support --compression-from
         if not self.recompress and not target.recreate_rechunkify:
             for chunk_id, size, csize in item.chunks:
                 self.cache.chunk_incref(chunk_id, target.stats)
             return item.chunks
         new_chunks = self.process_partial_chunks(target)
         chunk_iterator = self.create_chunk_iterator(archive, target, item)
         consume(chunk_iterator, len(new_chunks))
+        compress = self.compression_decider1.decide(item.path)
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: hmm, item.path - does it have a leading "/"?

Comment (enkore) [inline src/borg/archive.py]: It never has

Comment (ThomasWaldmann) [inline src/borg/archive.py]: if patterns in the compression config have leading slashes, then they won't ever match, right?

Comment (enkore) [inline src/borg/archive.py]: I guess. Kinda depends on how the patterns exactly look. Btw. the patterns are matched to paths relative to how they are specified on the command line in borg-create?

Comment (ThomasWaldmann) [inline src/borg/archive.py]: Someone could have a pattern on e.g. /home/user/*.

Comment (enkore) [inline src/borg/archive.py]: I see. I also don't really see how to fix it, the information whether a given path was absolute or relative is thrown away for good.

```
borg create testrepo::1235 /home/mabe/Projekte/_oss/borg/src --debug
using builtin fallback logging configuration
29 self tests completed in 0.13 seconds
Reading files cache ...
/home/mabe/Projekte/_oss/borg/src/borg/__pycache__/upgrader.cpython-35.pyc -> compression none
```

```
borg create testrepo::1234 src --debug
using builtin fallback logging configuration
29 self tests completed in 0.14 seconds
Reading files cache ...
src/borg/chunker.cpython-35m-x86_64-linux-gnu.so -> compression none
Upgrading from v1 hints.159
Upgrade to v2 hints complete
src/borg/hashindex.c -> compression none
```

Comment (enkore) [inline src/borg/archive.py]: conclusion: path in archives are always relative; that absolute patterns won't match is therefore expected.

Comment (ThomasWaldmann) [inline src/borg/archive.py]: expected for us. maybe the user does not expect this?

Comment (enkore) [inline src/borg/archive.py]: Since this also applies to all other patterns I documented it for now (5facc3a).

**Comment by enkore** [comment]
merge?

File: src/borg/archiver.py [hunk 2]
```diff
@@ -2032,6 +2039,10 @@ def build_parser(self, prog=None):
                                    help='keep tag files of excluded caches/directories')
 
         archive_group = subparser.add_argument_group('Archive options')
+        archive_group.add_argument('--target', dest='target', metavar='ARCHIVE', default=None,
+                                   type=archivename_validator(),
+                                   help='place archive at <target>, don\'t replace existing archive '
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: the word "place" here might suggest you can place it elsewhere (not in same repo).
also <target> is the option name, the value will show up as ARCHIVE.

maybe:
'create a new archive with the name ARCHIVE, do not replace existing archive'

File: src/borg/compress.pyx [hunk 3]
```diff
@@ -186,10 +188,14 @@ class Compressor:
         return self.compressor.compress(data)
 
     def decompress(self, data):
+        return self.detect(data)(**self.params).decompress(data)
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: maybe 2 lines, like
```
compressor_cls = ...
return compressor_cls(...
```

**Comment by ThomasWaldmann** [comment]
2 minor nitpicks, otherwise lgtm.

**Comment by enkore** [comment]
done & rebased on current master

**Pull request merged**